### PR TITLE
Unpacking code revised and provisionally tested

### PR DESCRIPTION
**Pull request by kenoir**
In order to implement the unpacking service we need to be able to unpack `tar.gz` files. This change updates an `Unpack` object to demonstrate this and test it.

### CODE REVIEW

File: bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/Unpack.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+object Unpack extends Logging {
+
+  def apply(inputStream: InputStream)(
+    out: ArchiveEntry => OutputStream
+  ): Stream[ArchiveEntry] = {
+
+    val readArchive = new ReadArchive(inputStream)
+
+    def entries(stream: InputStream): Stream[ArchiveEntry] = {
```
Comment (kenoir) [inline bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/Unpack.scala]: Tested this locally with a 5GB archive and it does not cause memory issues.

**Comment by alexwlchan** [comment]
I’ll have a look at this properly on Monday.

File: bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/Unpack.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+import scala.util.{Failure, Success, Try}
+
+object Unpack extends Logging {
+
+  def apply(inputStream: InputStream)(
+    out: ArchiveEntry => OutputStream
+  ): Stream[ArchiveEntry] = {
+
+    val readArchive = new ReadArchive(inputStream)
```
Comment (alexwlchan) [inline bag_unpacker/src/main/scala/uk/ac/wellcome/platform/archive/bagunpacker/Unpack.scala]: If I’ve read it correctly, this class can throw if passed an invalid archive - that feels odd for an `apply()` method.

File: bag_unpacker/src/test/scala/uk/ac/wellcome/platform/archive/bagunpacker/Archive.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  private val compressorOutputStream =
+    compress(outputStream)
+
+  private val archiveOutputStream = pack(
+    compressorOutputStream
+  )
+
+  def finish() = synchronized {
+    Future.fromTry(Try {
```
Comment (alexwlchan) [inline bag_unpacker/src/test/scala/uk/ac/wellcome/platform/archive/bagunpacker/Archive.scala]: Why are we wrapping this in `Future.fromTry(Try {`? Am I missing something, or isn’t this equivalent to `Future {`?

File: bag_unpacker/src/test/scala/uk/ac/wellcome/platform/archive/bagunpacker/UnpackTest.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    with ScalaFutures
+    with Compress
+    with RandomThings {
+
+  it("should unpack a tar.gz file") {
+    val (archiveFile, files, expectedEntriesF) =
+      createArchive(
+        "tar",
+        "gz",
+        10
```
Comment (alexwlchan) [inline bag_unpacker/src/test/scala/uk/ac/wellcome/platform/archive/bagunpacker/UnpackTest.scala]: Keyword arguments!

**Pull request merged**